### ISSUE TITLE
It should not consume system reserved computational resources

### ISSUE THREAD

**Issue by everpeace** [awslabs/mountpoint-s3-csi-driver#82]
I understand this CSI driver launches mountpoint-s3(FUSE) process in host's systemd.

In normal kubernetes node, cluster administrator configured appropriate reserved computing resource for system components (non-kubernetes components running in host). (https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/).

Launching fuse processes in host's systemd makes it difficult to manage computational resources and its reservation in the node, doesn't it? Particularly, when many pods using this csi driver are scheduled on the node, the situation could be worse.

When using the csi driver in kubernetes, how to manage node's computational resource as cluster administrator?? Are there any recommendation??

**Comment by jjkr** [comment]
That is correct, the CSI driver will launch the mountpoint fuse process in the host's systemd under `system.slice`, so using the system-reserved configuration that you linked is the only option for resource limits on the fuse driver itself at this time. Giving more fine grained control here is an enhancement we are looking at, however there is not a specific timeline I can share right now.

If there is any additional detail you would like to share that can help inform the design and prioritization of this feature, we'd be happy to hear it.

**Comment by everpeace** [comment]
> If there is any additional detail you would like to share that can help inform the design and prioritization of this feature, we'd be happy to hear it.

As I posted [here](https://github.com/awslabs/mountpoint-s3/issues/150#issuecomment-1676808125), I think:

- FUSE process should be running in sidecar container in user pod because of computational resource management perspective
- But, as a platform engineer, it would be difficult to allow privilege for the fuse sidecars but only for CSI driver containers. So, fuse sidecar without privilege would be nice.

There are several OSS that support this:
- https://github.com/GoogleCloudPlatform/gcs-fuse-csi-driver
  - FUSE process runs in non-privilege sidecar container in user pod
  - Only csi driver daemonset pod requires privilege
  - CSI driver just `mount` and the passed opened descriptor to sidecar container
  - Sidecar container just runs gcsfuse against the received already opened file descriptor
- https://github.com/pfnet-research/meta-fuse-csi-plugin
  - Almost the same technique with gcs-fuse-csi-driver
  - Additionally, they support `fusermount3-proxy` approach that can support existing FUSE implementation using `libfuse3` without any modification. `fusermount3-proxy` delegates `mount` execution to CSI driver pod(privileged) via unix domain socket.

**Comment by jjkr** [comment]
Thank you for the input. We are aware of this alternative design option and the advantages it brings for resource management in particular. The use of sidecars as they exist today also comes with some drawbacks, mostly related to the fact that the sidecar starts in parallel with the application container, so the mount is potentially not available to applications at startup and is not available to init containers at all.

The new [native sidecars](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) feature looks like a promising solution here and is something we're looking at. So again thank you for the request and the input and we'll share updates as soon as we can.

**Comment by javydekoning** [comment]
I think running as a side-car would also help support a broad range of OS'es and potentially resolve #103 #154 and #168

**Comment by everpeace** [comment]
> I think running as a side-car 

As stated [here](https://github.com/awslabs/mountpoint-s3-csi-driver/issues/82#issuecomment-1830915362), normally running FUSE process in a container would require privilege permission. It would arise security concern from kubernetes cluster admin perspective. So, running as side-car without __any privilege__ would be great 🙇

**Comment by muddyfish** [comment]
Requires Mountpoint to have improved memory control: https://github.com/awslabs/mountpoint-s3/issues/987

**Comment by M4t7e** [comment]
Removing the systemd dependency would enable compatibility with specialized Kubernetes operating systems like Talos, which do not use systemd. Currently, this integration appears infeasible (see #213).

**Comment by unexge** [comment]
Starting with v2.0.0, the CSI Driver spawns Mountpoint processes in dedicated Kubernetes Pods instead of using systemd. 

I believe this issue is resolved with v2. See https://github.com/awslabs/mountpoint-s3-csi-driver/issues/504 for more details about v2, and upgrading to it.
